include/boost/json/impl/pointer.ipp

```diff
@@ -0,0 +1,203 @@
+//
+// Copyright (c) 2021 Dmitry Arkhipov (grisumbras@gmail.com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+// Official repository: https://github.com/boostorg/json
+//
+
+#ifndef BOOST_JSON_IMPL_POINTER_IPP
+#define BOOST_JSON_IMPL_POINTER_IPP
+
+#include <boost/json/pointer.hpp>
+
+BOOST_JSON_NS_BEGIN
+
+namespace detail {
+
+std::size_t
+parse_number_token(
+    string_view::const_iterator& b,
+    string_view::const_iterator e,
+    error_code& ec) noexcept
+{
+    if ( ( b == e ) || ( *b == '0' && ( b + 1 ) != e  && b[1] != '/' ) )
+    {
+        ec = error::token_not_number;
+        return {};
+    }
+
+    std::size_t result = 0;
+    for (; b != e; ++b)
+    {
+        char const c = *b;
+
+        if ( '/' == c)
+        {
+            break;
+        }
+
+        if ( c < '0' || c > '9' )
+        {
+            ec = error::token_not_number;
+            return {};
+        }
+
+        result = result * 10 + c - '0';
+    }
+    return result;
+}
+
+
+string_view
+parse_string_token(
+    string_view::const_iterator& b,
+    string_view::const_iterator e,
+    char* out,
+    char const* out_end,
+    error_code& ec) noexcept
+{
+    char* const start = out;
+
+    for (; b < e; ++b)
+    {
+        char const c = *b;
+        if ( '/' == c )
+        {
+            break;
+        }
+
+        if ( out == out_end )
+        {
+            ec = error::token_too_large;
+            goto parse_done;
+        }
+
+        if ( '~' == c )
+        {
+            if ( ++b == e )
+            {
+                ec = error::invalid_escape;
+                goto parse_done;
+            }
+
+            switch (*b)
+            {
+            case '0':
+                *out++ = '~';
+                break;
+            case '1':
+                *out++ = '/';
+                break;
+            default:
+                ec = error::invalid_escape;
+                goto parse_done;
+            }
+        }
+        else
+        {
+            *out++ = c;
+        }
+    }
+
+parse_done:
+    return string_view(start, out - start);
+}
+
+template <class Value>
+Value&
+at_pointer(Value& root, pointer const& ptr, error_code& ec) noexcept
+{
+    ec.clear();
+    string_view const ptr_str = ptr.string();
+    Value* result = &root;
+    auto cur = ptr_str.begin();
+    while ( ptr_str.end() != cur )
+    {
+        if ( *cur++ != '/' )
+        {
+            ec = error::missing_slash;
+            return root;
+        }
+
+        if ( auto object = result->if_object() )
+        {
+            char buffer[256];
```

Comment:
Hard coded buffer size? Is overflow possible?

Reply:
I'm going to control buffer size with a macro.

Reply:
Can't we do better than that?

Reply:
Using a local buffer allows to make this function `noexcept`.  Also, if we allocate dynamically, we'd have to provide some way for allocation customization and that would clutter the API.

Reply:
>  if we allocate dynamically, we'd have to provide some way for allocation customization and that would clutter the API

I don't necessarily agree. We could have a local buffer, and go to the heap if the buffer is insufficient. We don't have to allow customization of the allocation if the memory is allocated and destroyed in the same function (i.e. if it is temporary). It becomes an implementation detail. I have done this in other places, such as the `stream_parser` on a suspend.